spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/outbound/AbstractMqttMessageHandler.java

```diff
@@ -109,8 +198,49 @@ protected void incrementClientInstance() {
 	@Override
 	protected void onInit() throws Exception {
 		super.onInit();
+		if (this.topicProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.topicProcessor).setBeanFactory(getBeanFactory());
+		}
+		if (this.qosProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.qosProcessor).setBeanFactory(getBeanFactory());
+		}
+		if (this.retainedProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.retainedProcessor).setBeanFactory(getBeanFactory());
+		}
 		if (this.converter == null) {
-			this.converter = new DefaultPahoMessageConverter(this.defaultQos, this.defaultRetained);
+			DefaultPahoMessageConverter mConverter = new DefaultPahoMessageConverter(
+					m -> {
+						if (this.qosProcessor == null) {
+							return MqttMessageConverter.defaultQosFunction(this.defaultQos).apply(m);
+						}
+						else {
+							Integer result = this.qosProcessor.processMessage(m);
+							if (result == null) {
+								return this.defaultQos;
+							}
+							else {
+								return result;
+							}
+						}
+					},
```

Comment:
Yes; I got a bit wrapped around the axle with this wrt tangling between the adapter and converter; I will rethink it a bit.